src/hashset.h

```diff
@@ -0,0 +1,188 @@
+#ifndef HASHSET_H
+#define HASHSET_H
+
+/* Hash table implementation.
+ *
+ * This is a cache-friendly hash table implementation. For details about the
+ * implementation and documentation of functions, see comments in hashset.c.
+ *
+ * The elements in a hashset are of a user-defined type, but an element needs to
+ * contain a key. It can represent a key-value entry, or it can be just a key,
+ * if set semantics are desired.
+ *
+ * Terminology:
+ *
+ * hashset
+ *         An instance of the data structure.
+ *
+ * element
+ *         An element in the hashset. This may be of the same type as the key,
+ *         or a struct containing a key and other fields.
+ * key
+ *         The part of the element used for looking the element up in the hashset.
+ *         May be the entire element or a struct field within the element.
+ *
+ * type
+ *         A struct containing callbacks, such as hash function, key comparison
+ *         function and how to get the key in an element.
+ */
+
+#include "fmacros.h"
+#include <stddef.h>
+#include <stdint.h>
+#include <unistd.h>
+
+/* --- Opaque types --- */
+
+typedef struct hashset hashset;
+typedef struct hashsetBucket hashsetBucket;
+typedef struct hashsetStats hashsetStats;
+
+/* --- Non-opaque types --- */
+
+/* The hashsetType is a set of callbacks for a hashset. All callbacks are
+ * optional. With all callbacks omitted, the hashset is effectively a set of
+ * pointer-sized integers. */
+typedef struct {
+    /* If the type of an element is not the same as the type of a key used for
+     * lookup, this callback needs to return the key within an element. */
+    const void *(*elementGetKey)(const void *element);
```

Comment:
Semantically, it's not really a set. I couldn't find a better name for this though.

The distinction between a key and an element can't be ignored though. When looking up an element, you provide a key, not a complete element. If we would require that, then the caller would need to create an robj with an embedded key just to lookup another robj with an embedded key. That adds overhead.

Reply:
Today, most of the code already does that.  And if we have a way to create a simplified element on the stack, it could be initialized with an SDS pointer with essentially zero overhead.  And this can all be done on the application side without bleeding those concepts into the hashset.